### PR TITLE
Added request to generateStateFunction

### PR DESCRIPTION
**Pull request by rmiraballes**
Add the functionality to generate a state dynamic with the data on the request.
Following: https://github.com/fastify/fastify-oauth2/issues/40

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -289,6 +289,44 @@ t.test('options.callbackUriParams', t => {
...
+    credentials: {
+      client: {
+        id: 'my-client-id',
+        secret: 'my-secret'
+      },
+      auth: oauthPlugin.GITHUB_CONFIGURATION
+    },
+    startRedirectPath: '/login/github',
+    callbackUri: '/callback',
+    generateStateFunction: (request) => request.query.code,
```
Comment (Eomm) [inline test.js]:
```suggestion
    generateStateFunction: (request) => {
      t.ok(request, 'the request param has been set')
      return request.query.code
    },
```

File: test.js [hunk 2]
```diff
@@ -289,6 +289,44 @@ t.test('options.callbackUriParams', t => {
...
+    fastify.inject({
+      method: 'GET',
+      url: '/login/github',
+      query: { code: 'generated_code' }
+    }, function (err, responseStart) {
+      t.error(err)
+      t.equal(responseStart.statusCode, 302)
+      const matched = responseStart.headers.location.match(/https:\/\/github\.com\/login\/oauth\/authorize\?response_type=code&client_id=my-client-id&redirect_uri=%2Fcallback&scope=notifications&state=generated_code/)
+      t.ok(matched)
+      t.end()
```
Comment (Eomm) [inline test.js]: could you use `t.plan()` instead of `t.end()`?
This is needed to check that the suggested `t.ok` in `generateStateFunction` is set

**Comment by rmiraballes** [comment]
Hi @Eomm, I just addressed your comments. Thanks!

File: README.md [hunk 3]
```diff
@@ -78,6 +78,33 @@ fastify.register(oauthPlugin, {
 })
 ```
 
+## Set custom state
+
+You can set a custom function to generate the state. In this case, is required to provide the function `checkStateFunction`.
+
+```js
+  fastify.register(oauthPlugin, {
```
Comment (Eomm) [inline README.md]:
```suggestion
  const validStates = new Set()

  fastify.register(oauthPlugin, {
```

File: README.md [hunk 4]
```diff
@@ -78,6 +78,33 @@ fastify.register(oauthPlugin, {
...
+      },
+      auth: oauthPlugin.FACEBOOK_CONFIGURATION
+    },
+    // register a fastify url to start the redirect flow
+    startRedirectPath: '/login/facebook',
+    // facebook redirect here after the user login
+    callbackUri: 'http://localhost:3000/login/facebook/callback',
+    // custom function to generate the state
+    generateStateFunction: (request) => {
+      return request.query.customCode
```
Comment (Eomm) [inline README.md]:
```suggestion
      const state = request.query.customCode
      validStates.add(state)
      return state
```

File: README.md [hunk 5]
```diff
@@ -78,6 +78,33 @@ fastify.register(oauthPlugin, {
...
+    },
+    // custom function to check the state is valid
+    checkStateFunction: () => true,
```
Comment (Eomm) [inline README.md]:
```suggestion
    checkStateFunction: (returnedState, callback) => {
      if (validStates.has(returnedState)) {
        callback()
        return
      }
      callback(new Error('Invalid state'))
    }
```

File: README.md [hunk 6]
```diff
@@ -78,6 +78,33 @@ fastify.register(oauthPlugin, {
 })
 ```
 
+## Set custom state
+
+You can set a custom function to generate the state. In this case, is required to provide the function `checkStateFunction`.
```
Comment (Eomm) [inline README.md]:
```suggestion
The `generateStateFunction` accepts a function to generate the `state` parameter for the OAUTH flow. This function receives the Fastify's `request` object as parameter.
When you set it, it is required to provide the function `checkStateFunction` also in order to validate the states generated.
```

**Comment by rmiraballes** [comment]
I updated the readme with your suggested changes

**Comment by rmiraballes** [comment]
@Eomm Can this be merged or is needed more approvals?

**Comment by Eomm** [comment]
Yes, by policy there must be at least 2 approvals before merge and landing

**Pull request merged**